File: app/src/main/java/workshop1024/com/xproject/view/dialog/TypeChoiceDialog.java
package workshop1024.com.xproject.view.dialog;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.view.ContextThemeWrapper;

import java.util.ArrayList;

import workshop1024.com.xproject.R;
import workshop1024.com.xproject.model.publishertype.PublisherType;

/**
 * 单选对话框
 */
public class TypeChoiceDialog extends DialogFragment implements DialogInterface.OnClickListener {
    public static final String TITLE_STRING_KEY = "TitleString";
    public static final String SELECT_TYPE_KEY = "SelectType";
    public static final String SELECTED_INDEX_KEY = "SelectedIndex";

    //对话框标题
    private String mTitleString;
    //对话框选项
    private ArrayList<PublisherType> mPublisherTypeList;
    //对话框选项索引
    private int mSelectedIndex;

    private TypeChoiceDialogListener mTypeChoiceDialogListener;

    public TypeChoiceDialog() {
    }

    /**
     * 创建单选对话框实例
     *
     * @param selectStrings 选择字符串数组资源id
     * @param titleStringId 标题字符串资源id
     * @param singleItemList
     * @param selectedIndex 选择的字符串索引
     * @return 单选对话框实例
     */
    public static TypeChoiceDialog newInstance(int titleStringId, ArrayList<PublisherType> publisherTypes, int selectedIndex) {
        TypeChoiceDialog typeChoiceDialog = new TypeChoiceDialog();
        Bundle bundle = new Bundle();
        bundle.putInt(TITLE_STRING_KEY, titleStringId);
        bundle.putSerializable(SELECT_TYPE_KEY, publisherTypes);
        bundle.putInt(SELECTED_INDEX_KEY, selectedIndex);
        typeChoiceDialog.setArguments(bundle);
        return typeChoiceDialog;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mTypeChoiceDialogListener = (TypeChoiceDialogListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement TypeChoiceDialogListener");
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle bundle = getArguments();
        if (bundle != null) {
            mTitleString = getResources().getString(bundle.getInt(TITLE_STRING_KEY));
            mPublisherTypeList = (ArrayList<PublisherType>) bundle.getSerializable(SELECT_TYPE_KEY);
            mSelectedIndex = bundle.getInt(SELECTED_INDEX_KEY);
        }
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(getActivity(),
                R.style.xproject_alertdialog));
        builder.setTitle(mTitleString).setItems(getTypeNameStrings(), this);
        return builder.create();
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int i) {
        mSelectedIndex = i;
        mTypeChoiceDialogListener.onTypeChoiceDialogItemClick(this, mPublisherTypeList.get(mSelectedIndex));
    }

    /**
     * 获取去选择的类型名称字符串集合
     *
     * @return 类型名称字符串集合
     */
    public String[] getTypeNameStrings() {
        String[] nameStrings = new String[mPublisherTypeList.size()];
        for (int type_i = 0; type_i < mPublisherTypeList.size(); type_i++) {
            nameStrings[type_i] = mPublisherTypeList.get(type_i).getName();
        }
        return nameStrings;
    }

    /**
     * 类型选择对话框点击监听接口
     */
    public interface TypeChoiceDialogListener {
        /**
         * 类型选择对话框选项点击回调方法
         *
         * @param dialog     点击的对话框
         * @param publisherType 点击的类型
         */
        void onTypeChoiceDialogItemClick(DialogFragment dialog, PublisherType publisherType);
    }
}

File: main/src/main/java/workshop1024/com/xproject/main/view/dialog/TypeChoiceDialog.kt
package workshop1024.com.xproject.main.view.dialog

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.ContextThemeWrapper
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import workshop1024.com.xproject.main.R

import java.util.ArrayList

/**
 * 单选对话框
 */
class TypeChoiceDialog : DialogFragment(), DialogInterface.OnClickListener {

    //对话框标题
    private var mTitleString: String? = null
    //对话框选项
    private var mPublisherTypeList: ArrayList<workshop1024.com.xproject.main.model.publishertype.PublisherType>? = null
    //对话框选项索引
    private var mSelectedIndex: Int = 0

    private var mTypeChoiceDialogListener: TypeChoiceDialogListener? = null

    /**
     * 获取去选择的类型名称字符串集合
     *
     * @return 类型名称字符串集合
     */
    val typeNameStrings: Array<String?>
        get() {
            val nameStrings: Array<String?> = arrayOfNulls<String>(mPublisherTypeList!!.size)
            for (type_i in mPublisherTypeList!!.indices) {
                nameStrings[type_i] = mPublisherTypeList!![type_i].name
            }
            return nameStrings
        }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        try {
            mTypeChoiceDialogListener = context as TypeChoiceDialogListener?
        } catch (e: ClassCastException) {
            throw ClassCastException(context!!.toString() + " must implement TypeChoiceDialogListener")
        }

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val bundle = arguments
        if (bundle != null) {
            mTitleString = resources.getString(bundle.getInt(TITLE_STRING_KEY))
            mPublisherTypeList = bundle.getSerializable(SELECT_TYPE_KEY) as ArrayList<workshop1024.com.xproject.main.model.publishertype.PublisherType>?
            mSelectedIndex = bundle.getInt(SELECTED_INDEX_KEY)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(ContextThemeWrapper(activity,
                R.style.xproject_alertdialog))
        builder.setTitle(mTitleString).setItems(typeNameStrings, this)
        return builder.create()
    }

    override fun onClick(dialogInterface: DialogInterface, i: Int) {
        mSelectedIndex = i
        mTypeChoiceDialogListener!!.onTypeChoiceDialogItemClick(this, mPublisherTypeList!![mSelectedIndex])
    }

    /**
     * 类型选择对话框点击监听接口
     */
    interface TypeChoiceDialogListener {
        /**
         * 类型选择对话框选项点击回调方法
         *
         * @param dialog     点击的对话框
         * @param publisherType 点击的类型
         */
        fun onTypeChoiceDialogItemClick(dialog: DialogFragment, publisherType: workshop1024.com.xproject.main.model.publishertype.PublisherType)
    }

    companion object {
        const val TITLE_STRING_KEY = "TitleString"
        const val SELECT_TYPE_KEY = "SelectType"
        const val SELECTED_INDEX_KEY = "SelectedIndex"

        /**
         * 创建单选对话框实例
         *
         * @param selectStrings 选择字符串数组资源id
         * @param titleStringId 标题字符串资源id
         * @param singleItemList
         * @param selectedIndex 选择的字符串索引
         * @return 单选对话框实例
         */
        fun newInstance(titleStringId: Int, publisherTypes: ArrayList<workshop1024.com.xproject.main.model.publishertype.PublisherType>, selectedIndex: Int): TypeChoiceDialog {
            val typeChoiceDialog = TypeChoiceDialog()
            val bundle = Bundle()
            bundle.putInt(TITLE_STRING_KEY, titleStringId)
            bundle.putSerializable(SELECT_TYPE_KEY, publisherTypes)
            bundle.putInt(SELECTED_INDEX_KEY, selectedIndex)
            typeChoiceDialog.arguments = bundle
            return typeChoiceDialog
        }
    }
}
